File: LeetCode/Study Plan/Practice Problems/Medium/Validate Binary Search Tree/Java/validateBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return recur(root, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
    }

    private boolean recur(TreeNode root, Double lowerBound, Double upperBound) {
        if (root == null) return true;
        return (root.val > lowerBound) &&
                (root.val < upperBound) &&
                recur(root.left, lowerBound, (double)root.val) &&
                recur(root.right, (double)root.val, upperBound);
    }
}

File: LeetCode/Medium/Validate Binary Search Tree/C/validate_bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
typedef struct TreeNode TreeNode;

bool recur(TreeNode* root, double lower_bound, double upper_bound) {
    return (root == NULL) ||
        (
            ((double)root->val > lower_bound && (double)root->val < upper_bound) &&
            recur(root->left, lower_bound, (double)root->val) &&
            recur(root->right, (double)root->val, upper_bound)
        );
}

bool isValidBST(TreeNode* root){
    return recur(root, -INFINITY, INFINITY);
}
